File: app/src/test/java/org/metplus/curriculum/web/security/ApplicationLoginFilterTest.java
package org.metplus.curriculum.web.security;

import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.Suite;
import org.metplus.curriculum.services.LocalTokenService;
import org.metplus.curriculum.services.TokenService;
import org.metplus.curriculum.test.BeforeAfterInterface;
import org.metplus.curriculum.test.BeforeAfterRule;
import org.metplus.curriculum.useCases.UserTryToLogin;
import org.springframework.mock.web.MockHttpServletRequest;
import org.springframework.mock.web.MockHttpServletResponse;

import static org.junit.Assert.*;

@RunWith(Suite.class)
@Suite.SuiteClasses({ApplicationLoginFilterTest.LoginSuccessfull.class,
        ApplicationLoginFilterTest.LoginError.class})
public class ApplicationLoginFilterTest {
    public static class DefaultLoginTest implements BeforeAfterInterface {
        @Rule
        public BeforeAfterRule beforeAfter = new BeforeAfterRule(this);
        protected UserTryToLogin useCase;
        protected TokenService tokenService;
        protected ApplicationLoginFilter filter;
        protected String validToken;
        @Override
        public void after() {
        }

        @Override
        public void before() {
            useCase = new UserTryToLogin("username", "password");
            tokenService = new LocalTokenService();
            validToken = tokenService.generateToken("1.1.1.1");
            filter = new ApplicationLoginFilter(useCase, tokenService);

        }
    }

    public static class LoginError extends DefaultLoginTest {
        @Test
        public void invalidCredentials_shouldReturnAuthenticationError() throws Exception {
            MockHttpServletRequest request = createRequest("1.1.1.1", "u", "p");
            MockHttpServletResponse response = new MockHttpServletResponse();
            assertFalse(filter.preHandle(request, response, filter));
            assertEquals(401, response.getStatus());
        }

        @Test
        public void noCredentialsPresentCredentials_shouldReturnAuthenticationError() throws Exception {
            MockHttpServletRequest request = new MockHttpServletRequest();
            MockHttpServletResponse response = new MockHttpServletResponse();
            assertFalse(filter.preHandle(request, response, filter));
            assertEquals(401, response.getStatus());
        }
    }

    public static class LoginSuccessfull extends DefaultLoginTest {

        @Test
        public void usernameAndLoginCorrect_shouldReturnGeneratedToken() throws Exception {
            MockHttpServletRequest request = createRequest("1.1.1.1", "username", "password");
            MockHttpServletResponse response = new MockHttpServletResponse();
            assertFalse(filter.preHandle(request, response, filter));
            assertEquals(200, response.getStatus());
            assertEquals("{\"token\": \"" + validToken + "\"}", response.getContentAsString());
            assertEquals("application/json", response.getContentType());
        }
    }

    private static MockHttpServletRequest createRequest(String remoteAddr, String username, String password) {
        MockHttpServletRequest request = new MockHttpServletRequest( );
        request.setRemoteAddr(remoteAddr);
        request.addHeader("X-Auth-Username", username);
        request.addHeader("X-Auth-Password", password);
        return request;
    }
}

File: web/src/test/kotlin/org/metplus/cruncher/web/security/filter/ApplicationLoginFilterTest.kt
package org.metplus.cruncher.web.security.filter

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.metplus.cruncher.web.security.filters.ApplicationLoginFilter
import org.metplus.cruncher.web.security.services.LocalTokenService
import org.metplus.cruncher.web.security.services.TokenService
import org.metplus.cruncher.web.security.useCases.UserTryToLogin
import org.springframework.mock.web.MockHttpServletRequest
import org.springframework.mock.web.MockHttpServletResponse

class ApplicationLoginFilterTest {
    open class DefaultLoginTest {
        protected lateinit var useCase: UserTryToLogin
        protected lateinit var tokenService: TokenService
        protected lateinit var filter: ApplicationLoginFilter
        protected lateinit var validToken: String

        @BeforeEach
        fun before() {
            useCase = UserTryToLogin("username", "password")
            tokenService = LocalTokenService()
            validToken = tokenService.generateToken("1.1.1.1")
            filter = ApplicationLoginFilter(useCase, tokenService)

        }
    }

    class LoginError : DefaultLoginTest() {
        @Test
        @Throws(Exception::class)
        fun invalidCredentials_shouldReturnAuthenticationError() {
            val request = createRequest("1.1.1.1", "u", "p")
            val response = MockHttpServletResponse()
            assertFalse(filter.preHandle(request, response, filter))
            assertThat(response.status).isEqualTo(401)
        }

        @Test
        @Throws(Exception::class)
        fun noCredentialsPresentCredentials_shouldReturnAuthenticationError() {
            val request = MockHttpServletRequest()
            val response = MockHttpServletResponse()
            assertFalse(filter.preHandle(request, response, filter))
            assertThat(response.status).isEqualTo(401)
        }
    }

    class LoginSuccessfull : DefaultLoginTest() {

        @Test
        @Throws(Exception::class)
        fun usernameAndLoginCorrect_shouldReturnGeneratedToken() {
            val request = createRequest("1.1.1.1", "username", "password")
            val response = MockHttpServletResponse()
            assertFalse(filter.preHandle(request, response, filter))
            assertThat(response.status).isEqualTo(200)
            assertThat(response.contentAsString).isEqualTo("{\"token\": \"$validToken\"}")
            assertThat(response.contentType).isEqualTo("application/json")
        }
    }
}

private fun createRequest(remoteAddr: String, username: String, password: String): MockHttpServletRequest {
    val request = MockHttpServletRequest()
    request.remoteAddr = remoteAddr
    request.addHeader("X-Auth-Username", username)
    request.addHeader("X-Auth-Password", password)
    return request
}